examples/controls_example/public/react_controls/control_group/control_group_unsaved_changes_api.ts

```diff
@@ -68,12 +70,22 @@ export function initializeControlGroupUnsavedChanges(
           return Object.keys(unsavedChanges).length ? unsavedChanges : undefined;
         })
       ),
-      resetUnsavedChanges: () => {
+      asyncResetUnsavedChanges: async () => {
         controlGroupUnsavedChanges.api.resetUnsavedChanges();
+
+        const filtersReadyPromises: Array<Promise<void>> = [];
         Object.values(children$.value).forEach((controlApi) => {
           if (apiPublishesUnsavedChanges(controlApi)) controlApi.resetUnsavedChanges();
+          if (apiPublishesAsyncFilters(controlApi)) {
+            filtersReadyPromises.push(controlApi.untilFiltersReady());
+          }
         });
+
+        await Promise.all(filtersReadyPromises);
+        applySelections();
```

Comment:
We should probably only call `applySelections` **if** auto-apply is disabled - otherwise, we are publishing twice when it is enabled due to the `filter$` subscription in the control group.

Reply:
resolve with https://github.com/elastic/kibana/pull/189830/commits/f1a4c63ed64c0034fbb3bc3dc0c3146502eea390

Reply:
I added an if statement around `applySelections`. `applySelections` will not publish filters if there are no changes. See the below implemenation

```
function applySelections() {
    if (!deepEqual(filters$.value, unpublishedFilters$.value)) {
      filters$.next(unpublishedFilters$.value);
    }
    if (!deepEqual(timeslice$.value, unpublishedTimeslice$.value)) {
      timeslice$.next(unpublishedTimeslice$.value);
    }
  }
```

Reply:
Ah, good point - I personally still think we should keep the `applySelections` in an `if` (there is more overhead to calling `applySelections` twice than there is to checking the value of `applySelections`). But ultimately I'm fine either way, since the at least the `filters$` subject won't emit twice :+1: